src/Workspaces/Core/Portable/Remote/RemoteHostSessionHelpers.cs

```diff
@@ -0,0 +1,292 @@
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Execution;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Remote
+{
+    /// <summary>
+    /// Helper type for common session case
+    /// 
+    /// This will tie <see cref="Solution"/> and <see cref="RemoteHostClient.Session"/>'s lifetime together
+    /// so that one can handle those more easily
+    /// </summary>
+    internal sealed class SolutionAndSessionHolder : IDisposable
+    {
+        private readonly RemoteHostClient.Session _session;
+        private readonly PinnedRemotableDataScope _scope;
+        private readonly CancellationToken _cancellationToken;
+
+        public static async Task<SolutionAndSessionHolder> CreateAsync(RemoteHostClient.Session session, PinnedRemotableDataScope scope, CancellationToken cancellationToken)
+        {
+            var sessionWithSolution = new SolutionAndSessionHolder(session, scope, cancellationToken);
+
+            try
+            {
+                await session.RegisterPinnedRemotableDataScopeAsync(scope).ConfigureAwait(false);
+                return sessionWithSolution;
+            }
+            catch
+            {
+                sessionWithSolution.Dispose();
+
+                // we only expect this to happen on cancellation. otherwise, rethrow
+                cancellationToken.ThrowIfCancellationRequested();
+                throw;
+            }
+        }
+
+        private SolutionAndSessionHolder(RemoteHostClient.Session session, PinnedRemotableDataScope scope, CancellationToken cancellationToken)
+        {
+            _session = session;
+            _scope = scope;
+            _cancellationToken = cancellationToken;
+        }
+
+        public void AddAdditionalAssets(CustomAsset asset)
+        {
+            _scope.AddAdditionalAsset(asset, _cancellationToken);
+        }
+
+        public void Dispose()
+        {
+            _scope.Dispose();
+            _session.Dispose();
+        }
+
+        public Task InvokeAsync(string targetName, params object[] arguments) =>
+            _session.InvokeAsync(targetName, arguments);
+        public Task<T> InvokeAsync<T>(string targetName, params object[] arguments) =>
+            _session.InvokeAsync<T>(targetName, arguments);
+        public Task InvokeAsync(string targetName, IEnumerable<object> arguments, Func<Stream, CancellationToken, Task> funcWithDirectStreamAsync) =>
+            _session.InvokeAsync(targetName, arguments, funcWithDirectStreamAsync);
+        public Task<T> InvokeAsync<T>(string targetName, IEnumerable<object> arguments, Func<Stream, CancellationToken, Task<T>> funcWithDirectStreamAsync) =>
+            _session.InvokeAsync<T>(targetName, arguments, funcWithDirectStreamAsync);
```

Comment:
Note: if you *do* want to use containment instead of inheritance, then i ask that you just expose everything that features need through an ISession interface instead of a Session concrete type.  Then SolutionAndSessionHolder can be an internal class that implements ISession.  

My primary concern here is that the *pseudo-public* API (i.e. the API that features use) is appropriate designed, without unnecessary concepts that they shouldn't have to be aware of.  If you go this route, you can still have your different implementation classes (if you find them helpful in the engine layer), but all features only ever need to interact with a simple interface that does not expose disparate types that they don't need to know about.